### ISSUE TITLE
Using CSS layers breaks rules which utilize variables

### ISSUE THREAD

**Issue by a-type** [unocss/unocss#2892]
### UnoCSS version

0.53.5

### Describe the bug

When applying rules like `shadow-*` inside a `layer-xxx:` CSS layer, the variables utilized to control shorthand properties are all overridden by the default variable declarations applied to `*`.

<img width="408" alt="image" src="https://github.com/unocss/unocss/assets/2829772/0ca01945-ae20-490d-8986-966c6043a91c">

While I understand this is technically working as designed (the rule was generated in the layer as specified, and any non-layer CSS like those variable declarations will override layer rules) it's unexpected from the perspective of usage, and leads to a lot of implicit behavior causing silent CSS failures like the one pictured only for particular rules.

My goal in using CSS layers here was to set up some default built-in styles for reusable components which can be easily overridden at the point of usage without worrying about specificity-- so styles for "Button" will all be in a `components` layer, button variants will be in `variants`, and then all user-applied classes will beat both of those. It seemed effective and even elegant, but now I can't apply box-shadows or transforms or a number of other rules as part of this paradigm.

### Reproduction

https://unocss.dev/play/?html=DwIwrgLhD2B2AEBjANgQwM7oLwCI0E8BTAJwFoAPALnQAtUATaAd3gJIursadNQFsQJHAD4AKoXQRgAenBQ4wgFChIMBCgzYctBs3g7uvAULESps1QqA&config=JYWwDg9gTgLgBAbwFBzgEwKYDNgDsMDCEuOA5gDQpxhQYDOGMAgjDFMAEYCuMwWAnpVQ16jAJIBjYnSHVaDGAFVcESgF84WKBBBwA5FxUS6dPUiQYAHpFjpsAQy4AbeJhz4iJYKQAUyVFDO9ABccADaVKhhehJcdDA6ALSBThh65IhwUk7QoXq0aHpwagC6VCWydAAW0DCxMHSh-qj6sfFJ1bX1enkwVjCJTqRwfZYD0Pa4pBhwNQBuGFDBowN99k5mqGqyIgqN4ZFyokoqPgCUssLyjCxsnDx8-OeXRwqS0n6HqHQS6xihAEYAHQAJheqAkaFweSqrDAjQA9Aj6CAgdUEZsWmoLuUkNikEA&css=Q&options=N4XyA

### System Info

Windows 11, Chrome 114, React

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by a-type** [comment]
After thinking about this, I guess what I really need to do is move the base variable declarations from `preset-mini` to a layer which is ordered with less precedence than my other layers. Makes sense that this is not something Uno can really do without understanding my layer structure.

Is it too picky to have this be configurable in the built-in presets? Seems like I should just fork and maintain my own if I need this, but building from an existing preset is really convenient.

Feel free to close if this seems like a wontfix. Once nice thing about building all functionality in presets/plugins is I do have a userland way to fix the behavior, even if it is tedious.

**Comment by a-type** [comment]
I thought a bit harder and I've found a solution that works for me. Perhaps someone else will need to solve this too.

I turned off preflights in my Uno preset, then recreated them in a new CSS layer, following the preset-mini source code...

```
{
	layer: 'preflights',
	getCSS: (ctx: PreflightContext<any>) => {
		if (ctx.theme.preflightBase) {
			const css = entriesToCss(Object.entries(ctx.theme.preflightBase));
			const roots = toArray(
				ctx.theme.preflightRoot ?? ['*,::before,::after', '::backdrop'],
			);
			return roots
				.map((root) => `@layer preflightBase{${root}{${css}}}`)
				.join('');
		}
	},
} as any,
```

the `any` cast is just because the theme situation is more complicated for my custom preset and I didn't want to figure out typings.

The relevant part is adding `@layer preflightBase{}` around the original `${root}{${css}}` bit.

Then I added this new layer to my existing layer ordering in my own preflight

```css
@layer preflightBase, components, variants, utilities;
```

And that seems to have worked great, without having to ditch my Uno preset. The base CSS vars are all defined within their own low-ranking layer, so other layered rules take precedence and can override them correctly.
